### PR TITLE
[BugFix] conditional expression process const columns error

### PR DESCRIPTION
**Pull request by fzhedu**
fix conditional expression process const columns error, include if, ifnull, nullif, coalesce

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

File: be/src/exprs/condition_expr.cpp [hunk 1]
```diff
@@ -186,8 +192,12 @@ class VectorizedNullIfExpr : public Expr {
...
     }
 
-    ColumnPtr _evaluate_complex(const Columns& columns) {
-        auto num_rows = columns[0]->size();
+    ColumnPtr _evaluate_complex(const Columns& inputs) {
+        auto num_rows = inputs[0]->size();
+        Columns columns;
+        for (const auto& col : inputs) {
+            columns.push_back(ColumnHelper::unpack_and_duplicate_const_column(num_rows, col));
+        }
```
Comment (zombee0) [inline be/src/exprs/condition_expr.cpp:198]: there is no need to deal const_column for nullifexpr and coalesceexpr, they will use other branch.

File: be/src/exprs/condition_expr.cpp [hunk 2]
```diff
@@ -436,18 +448,22 @@ class VectorizedCoalesceExpr : public Expr {
         return builder.build(ColumnHelper::is_all_const(columns));
     }
 
-    StatusOr<ColumnPtr> _evaluate_complex(const Columns& columns) { // without only-null columns
-        int size = columns[0]->size();
+    StatusOr<ColumnPtr> _evaluate_complex(const Columns& inputs) { // without only-null columns
```
Comment (zombee0) [inline be/src/exprs/condition_expr.cpp:451]: there is no need to deal const_column for coalesceexpr, they will use other branch.

Comment (zombee0) [reply be/src/exprs/condition_expr.cpp:451]: Got it

File: be/src/exprs/condition_expr.cpp [hunk 3]
```diff
@@ -198,7 +206,7 @@ class VectorizedNullIfExpr : public Expr {
         }
         for (int row = 0; row < num_rows; ++row) {
             if ((right_nulls == nullptr || !right_nulls->get_data()[row]) &&
-                columns[0]->equals(row, *right_data, row)) {
+                columns[0]->equals(row, *right_data, row, false) == 1) {
                 res->append_nulls(1);
```
Comment (zombee0) [inline be/src/exprs/condition_expr.cpp:210]: Could you confirm this

**Review by Seaven — DISMISSED** [review]

**Review by zombee0 — APPROVED** [review]
LGTM

**Review by silverbullet233 — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**